node/runtime/src/impls.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Some configurable implementations as associated type for the substrate runtime.
+
+use node_primitives::Balance;
+use runtime_primitives::weights::{Weight, MAX_TRANSACTIONS_WEIGHT, IDEAL_TRANSACTIONS_WEIGHT};
+use runtime_primitives::traits::{Convert, Zero};
+use crate::{Runtime, Perbill, Balances};
+
+/// Struct that handles the conversion of Balance -> `u64`. This is used for staking's election
+/// calculation.
+pub struct CurrencyToVoteHandler;
+
+impl CurrencyToVoteHandler {
+	fn factor() -> Balance { (Balances::total_issuance() / u64::max_value() as Balance).max(1) }
+}
+
+impl Convert<Balance, u64> for CurrencyToVoteHandler {
+	fn convert(x: Balance) -> u64 { (x / Self::factor()) as u64 }
+}
+
+impl Convert<u128, Balance> for CurrencyToVoteHandler {
+	fn convert(x: u128) -> Balance { x * Self::factor() }
+}
+
+/// Struct used to convert from a transaction weight into the actual fee value.
+/// This is used in the balances module.
+///
+/// This assumes that weight is a numeric value in the u32 range.
+///
+/// Formula:
+///   diff = (ideal_weight - current_block_weight)
+///   v = 0.00004
+///   fee = weight * (1 + (v . diff) + v^2 (v . d)^2 / 2)
+///
+/// https://research.web3.foundation/en/latest/polkadot/Token%20Economics/#relay-chain-transaction-fees
+pub struct WeightToFeeHandler;
```

Comment:
This needs a bit more documentation. I followed the link and as far as I understand it, the logic contained here is for a dynamic system that allows a single-step fee change once per block. The documentation here seems to imply that this logic is used whenever an individual transaction's fee must be calculated from its weight.

Reply:
I checked and indeed it seems this logic is used by `make_payment`. Yet reading the original research it seems clear that this logic is only used once per block in order to recalibrate the fee multiplier. Something is amiss.

Reply:
You're right this was misinterpreted by me. Won't probably be that hard to fix. One new storage item in system `NextFeeMultiplier: Perbill`, which gets updated for the next block at the end of each block (with the same update logic).

Reply:
@Kianenigma and I just discussed this. At the moment, the fee multiplier is updated per transaction within a block, but we are in agreement that this doesn't make much sense. The correct approach should be to update the fee multiplier per block according to the average block saturation relative to the current block's saturation in the context of some ideal target block saturation. The correct implementation will add a `new_fee_multiplier` storage item that will be updated at the end of every block according to changes in the block saturation history.

Reply:
This can be reviewed again.